subprojects/execution/src/main/java/org/gradle/internal/execution/steps/StoreSnapshotsStep.java

```diff
@@ -23,38 +23,31 @@
 import org.gradle.internal.execution.Step;
 import org.gradle.internal.execution.UnitOfWork;
 import org.gradle.internal.execution.history.AfterPreviousExecutionState;
-import org.gradle.internal.execution.history.OutputFilesRepository;
 import org.gradle.internal.execution.history.changes.OutputFileChanges;
 import org.gradle.internal.fingerprint.CurrentFileCollectionFingerprint;
 import org.gradle.internal.fingerprint.FileCollectionFingerprint;
 
 import java.util.Optional;
 
 public class StoreSnapshotsStep<C extends IncrementalContext> implements Step<C, CurrentSnapshotResult> {
-    private final OutputFilesRepository outputFilesRepository;
     private final Step<? super C, ? extends CurrentSnapshotResult> delegate;
 
     public StoreSnapshotsStep(
-        OutputFilesRepository outputFilesRepository,
         Step<? super C, ? extends CurrentSnapshotResult> delegate
     ) {
-        this.outputFilesRepository = outputFilesRepository;
         this.delegate = delegate;
     }
 
     @Override
-    // TODO Return a simple Result (that includes the origin metadata) here
     public CurrentSnapshotResult execute(C context) {
         CurrentSnapshotResult result = delegate.execute(context);
         ImmutableSortedMap<String, CurrentFileCollectionFingerprint> finalOutputs = result.getFinalOutputs();
         context.getBeforeExecutionState().ifPresent(beforeExecutionState -> {
             boolean successful = result.getOutcome().isSuccessful();
-            Optional<AfterPreviousExecutionState> afterPreviousExecutionState = context.getAfterPreviousExecutionState();
             // Only persist history if there was no failure, or some output files have been changed
```

Comment:
I don't think that comment adds anything anymore. We should remove it. If we want to add a comment, then we should say why we only persist the history if there was no failure, or some output files have been changed.

Reply:
Good point, changed the comment.

Reply:
Reading that comment... should we be storing the snapshots if the task was successful but there were no updates to the outputs? I mean, we now overwrite the origin metadata with that of the current build, but the content was already there, produced by a previous one. Maybe what we should also skip the store if the previous execution and the current one were both successful?

Reply:
I think we should. For incremental compilation for example, the task can decide that there is nothing to compile, even though the inputs changed. So we should store the fact that the last execution had different inputs than the one before.

The only way to have inputs and outputs the same before and after task execution without the task being up-to-date is when we are in `re-run` mode, right? Then then intention is to store the new result in the task history I think.

WDYT?

Reply:
Mhm, you're right.